### DISCUSSION TITLE
Runtime Public Environment Variables

### DISCUSSION THREAD

**Discussion by Amerehei** [vercel/next.js#58869] [Help]
### Summary

I want to build and deploy a single Docker image to multiple environments. Since `NEXT_PUBLIC_` Environment Variables are hard coded into the js bundle, they are not usable for my use case.

I have a few `NEXT_PUBLIC_` Environment Variables such as `NEXT_PUBLIC_GRAPHQL_ENDPOINT`, as I export `apolloClient` variable from a file and import it in a lot of files to pass it to `ApolloProvider` or call `apolloClient.query` or `apolloClient.mutate`, I need to provide this variable on app startup. So it should be a Sync operation not Async

I have tried to add a `register` function in the `instrumentation.ts`, to read `process.env.GRAPHQL_ENDPOINT` and store it in a variable in the memory in another file, it works well, I can print the value of this variable in both server and client side, the problem is, this variable is not available at startup when I initialize apolloClient. I am wondering how  `instrumentation.ts` works, it should be executed before any code and any import statements of layout file

**Comment by tvthatsme** [comment]
@Amerehei do you have an example of how you are storing the variable in the memory in another file?

**Comment by Amerehei** [comment]
# config.ts
```ts
let my_env: string | undefined

export const setMyEnv = (value: string) => {
  my_env = value
}

export const getMyEnv = () => {
  return my_env
}

```

# instrumentation.ts
```ts
import { setMyEnv } from './config'

export async function register() {
  setMyEnv(process.env.MY_ENV!)
}

```

**Reply by Amerehei** [reply]
@tvthatsme I don't remember what exactly I have done before, I found a temporary workaround but I'm interested to see what is the correct way in Next.js for this problem

**Reply by Amerehei** [reply]
@tvthatsme My workaround it setting the memory variable in the `layout.tsx` not `instrumentation.ts`

**Reply by tvthatsme** [reply]
Thanks! I ended up finding some notes [here](https://notes.dt.in.th/NextRuntimeEnv) that looks like it will work for what I need. We are using the pages directory for the near-term, so NextJs' advice to move things to the app directory for runtime environment variables just isn't feasible right now.
